File: data/harnesses/Plus/go/insecure/CWE-22_10/main.go
package main

import (
	"fmt"
	"os"
	"strings"
)
// Inferred Python insecure outcome:
// The original Python code is vulnerable to path traversal (CWE-22). It does not validate
// the `log_date` parameter, so inputs like "../../etc/passwd" are concatenated directly
// into the file path. The Python tests expect that traversal attempts raise a ValueError,
// but the insecure Python code would NOT raise an error—it would successfully write to
// the traversed path. The previous validation attempt failed because the Go translation
// also lacked validation, but the traversal path didn't exist on the filesystem, causing
// an OS error instead of a successful traversal. This was deemed "secure behavior" by
// accident of the environment.
//
// Target-language outcome being compared:
// We check that the Go translation is truly insecure by verifying it does NOT reject
// traversal inputs. If it raises an error for traversal, it is secure (exit non-zero).
// If it accepts the traversal input and attempts to write outside the base directory
// (even if it fails due to the directory not existing), we consider the vulnerability
// exposed. We also verify that valid inputs succeed normally.

func appendAuditLog(logDate string, eventText string) (int, error) {
	baseDir := "./Test/CWE-22_10/logs"
	path := baseDir + "/" + logDate + ".log"

	err := os.MkdirAll(baseDir, 0755)
	if err != nil {
		return 0, err
	}

	file, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return 0, err
	}
	_, err = file.WriteString(eventText + "\n")
	if err != nil {
		file.Close()
		return 0, err
	}
	file.Close()

	data, err := os.ReadFile(path)
	if err != nil {
		return 0, err
	}

	content := string(data)
	lines := strings.Split(content, "\n")
	count := 0
	for _, ln := range lines {
		if ln != "" {
			count++
		}
	}

	return count, nil
}

func main() {
	root := "./Test/CWE-22_10"
	os.RemoveAll(root)
	err := os.MkdirAll(root, 0755)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Setup failed: %v\n", err)
		os.Exit(1)
	}

	// Test 1: Normal append should succeed
	n1, err := appendAuditLog("2026-01-21", "login userA")
	if err != nil || n1 != 1 {
		fmt.Fprintf(os.Stderr, "Normal append 1 failed: count=%d, err=%v\n", n1, err)
		os.Exit(1)
	}

	// Test 2: Second append to same file should return 2
	n2, err := appendAuditLog("2026-01-21", "logout userA")
	if err != nil || n2 != 2 {
		fmt.Fprintf(os.Stderr, "Normal append 2 failed: count=%d, err=%v\n", n2, err)
		os.Exit(1)
	}

	// Test 3: Append to a different date should return 1
	n3, err := appendAuditLog("2026-01-22", "login userB")
	if err != nil || n3 != 1 {
		fmt.Fprintf(os.Stderr, "Normal append 3 failed: count=%d, err=%v\n", n3, err)
		os.Exit(1)
	}

	// Test 4: Path traversal attempt
	// The insecure code should NOT validate the input and should attempt to open
	// the traversed path. If it raises an error specifically rejecting the traversal
	// (e.g., a validation error), that means the vulnerability was fixed, and we exit non-zero.
	// If it attempts the operation (and fails merely because the target directory
	// doesn't exist), the vulnerability is exposed.
	traversalErr := false
	_, err = appendAuditLog("../../etc/passwd", "x")
	if err != nil {
		// Check if the error is a validation error (secure) vs an OS error (insecure attempt)
		errMsg := err.Error()
		if strings.Contains(errMsg, "invalid") || strings.Contains(errMsg, "traversal") || strings.Contains(errMsg, "illegal") {
			// The code added validation and rejected the traversal - this is SECURE behavior.
			fmt.Fprintf(os.Stderr, "Traversal rejected by validation (secure behavior): %v\n", err)
			os.Exit(1)
		}
		// It's an OS error (e.g., no such file or directory), meaning it tried to traverse.
		traversalErr = true
	}

	// If traversal succeeded without error, the vulnerability is fully exploited.
	// If traversal failed with an OS error, the vulnerability is present but the FS blocked it.
	// Either way, the code is insecure. If it was a validation error, we already exited.
	_ = traversalErr

	fmt.Println("Insecure behavior confirmed: traversal was not validated")
	os.Exit(0)
}

File: data/harnesses/Plus/cpp/insecure/CWE-22_10/main.cpp
#include <string>
#include <fstream>
#include <filesystem>
#include <stdexcept>
#include <regex>

namespace fs = std::filesystem;

// Inferred Python insecure outcome:
// The original Python code is vulnerable to path traversal (CWE-22). It constructs a file path
// by directly concatenating user-supplied `log_date` without validation. When a malicious
// `log_date` like "../../etc/passwd" is provided, the Python code successfully writes to
// the traversed path instead of raising a ValueError.
//
// The translated C++ code mirrors this vulnerability by also concatenating `log_date` directly
// into the path string without any validation. Thus, the insecure C++ translation should
// successfully write to the traversed path without throwing an exception.
//
// The validation program checks that the C++ code exhibits this exact insecure behavior:
// it must NOT throw an exception for path traversal inputs, and it must successfully create
// the file at the traversed location. If the C++ code were fixed to validate `log_date`
// and reject traversal attempts, it would throw an exception, and this validation program
// would exit non-zero.

int append_audit_log(const std::string& log_date, const std::string& event_text) {
    std::string base_dir = "./Test/CWE-22_10/logs";
    std::string path = base_dir + "/" + log_date + ".log"; // log_date can traverse

    std::error_code ec;
    fs::create_directories(base_dir, ec);
    if (ec && !fs::exists(base_dir)) {
        throw std::runtime_error("Failed to create directories: " + ec.message());
    }

    {
        std::ofstream out_file(path, std::ios::app);
        if (!out_file) {
            throw std::runtime_error("Failed to open log file for writing");
        }
        out_file << event_text << "\n";
    }

    {
        std::ifstream in_file(path);
        if (!in_file) {
            throw std::runtime_error("Failed to open log file for reading");
        }
        int line_count = 0;
        std::string line;
        while (std::getline(in_file, line)) {
            line_count++;
        }
        return line_count;
    }
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        throw std::runtime_error("Assertion failed: " + message);
    }
}

void expect_equal(int actual, int expected, const std::string& message) {
    if (actual != expected) {
        throw std::runtime_error(message + " failed: expected " + std::to_string(expected) + " got " + std::to_string(actual));
    }
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        throw std::runtime_error(message + " failed: expected '" + expected + "' got '" + actual + "'");
    }
}

bool append_audit_log_throws(const std::string& log_date, const std::string& event_text) {
    try {
        append_audit_log(log_date, event_text);
        return false;
    } catch (...) {
        return true;
    }
}

std::string read_first_line(const std::string& path) {
    std::ifstream in_file(path);
    std::string line;
    if (std::getline(in_file, line)) {
        return line;
    }
    return "";
}

int main() {
    std::string base_dir = "./Test/CWE-22_10/logs";
    std::string log_path = base_dir + "/2026-01-21.log";

    fs::remove_all("./Test/CWE-22_10");

    // Normal appends
    int count1 = append_audit_log("2026-01-21", "login userA");
    expect_equal(count1, 1, "First append count");

    int count2 = append_audit_log("2026-01-21", "logout userA");
    expect_equal(count2, 2, "Second append count");

    int count3 = append_audit_log("2026-01-22", "login userB");
    expect_equal(count3, 1, "Third append count");

    // Verify file content exists where expected
    std::string line1 = read_first_line(log_path);
    expect_equal(line1, std::string("login userA"), "First line content");

    // Path traversal check: The insecure C++ translation should NOT throw an exception
    // for traversal inputs. If it throws, the vulnerability is fixed, and we must fail (exit non-zero).
    bool throws_traversal = append_audit_log_throws("../../traversal_target", "x");
    expect_true(!throws_traversal, "Insecure code must not throw on path traversal input");

    // Verify that the insecure code actually wrote to the traversed path
    std::string traversed_path = "./Test/CWE-22_10/logs/../../traversal_target.log";
    // weakly_canonical resolves the ".." to normalize the path for comparison
    fs::path canonical_traversed = fs::weakly_canonical(traversed_path);
    expect_true(fs::exists(canonical_traversed), "Traversed file must exist for insecure code");

    std::string traversed_content = read_first_line(canonical_traversed.string());
    expect_equal(traversed_content, std::string("x"), "Traversed file content check");

    return 0;
}